presto-main/src/main/java/com/facebook/presto/sql/analyzer/PredicateStitcher.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.sql.tree.AliasedRelation;
+import com.facebook.presto.sql.tree.AllColumns;
+import com.facebook.presto.sql.tree.AstVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.Join;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.Query;
+import com.facebook.presto.sql.tree.QueryBody;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Relation;
+import com.facebook.presto.sql.tree.Table;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.metadata.MetadataUtil.createQualifiedObjectName;
+import static com.facebook.presto.metadata.MetadataUtil.toSchemaTableName;
+import static com.facebook.presto.sql.QueryUtil.selectList;
+import static com.facebook.presto.sql.QueryUtil.subquery;
+
+public class PredicateStitcher
```

Comment:
This is just like `PredicatePushDown`. Feel free to check that class. It would be fairly complicated. That is the reason I gave the `LEFT JOIN` example, where predicates could not be pushed down through outer joins. This will be a larger discussion what query shapes we should support...

If we want to limit the query shape, this class is a perfect one. We can just write `visitNode(Node node, C context)` method here by failing any unexpected Node (e..g, `JOIN`)

Reply:
Yeah, we would need something to prevent not supported query plans. However this place might be too late in the execution. This may not even be called if materialized view is fully materialized. I had added a TODO at the first reference of materialized view (in statement analyzer) to validate the supported plans.